### ISSUE TITLE
Modify API Response Rule modifies the xhr.response type

### ISSUE THREAD

**Issue by mdpratt** [requestly/requestly#252]
## Summary

When attempting to use a "Modify API Response Rule" to return a custom object, it's changing the typeof xhr.response from `object` to `string`. This is causing downstream impacts where response parsing is failing. Is this intented?

Here is the standard response from our API:

![image](https://user-images.githubusercontent.com/982135/207246100-cc703df8-7a4a-459d-84d2-05b938e7a60f.png)

Here is the response after enabling Requestly's modify API response rule:

![image](https://user-images.githubusercontent.com/982135/207246124-c44c11ed-1d68-46b7-b832-5fcc3eeeacbb.png)

**Comment by sachinjain024** [comment]
Hey, @mdpratt Thanks a lot for creating a ticket here. 

Is it possible for you to share the page URL where we can test this out and also help us with the API URL or your rule? We will investigate this on priority.

**Comment by mdpratt** [comment]
Put together a sample using the TS Playground:
https://www.typescriptlang.org/play?ssl=3&ssc=87&pln=3&pc=15#code/PTAEEkDsGcBcCcCuBjWBLA9jAXAKBKIQIwB0oAwgDZrIDWoARAEqKQOgCGkAJqAJ4ZEoAO5pKlUNACmU0JSkAzWKHhToAByzSSAIwzc+nHirWJKytJFCwAFrMoYA5tHxhCAJjIA1NNDTKbWFh1aGwQDnV1ElUAR0Q1WEo+EkxgJHloYABiaBsOVW4AGV9YYCIANgB2AAYATndygFYAFgAOAGYmgFomKUdVaD8sHqlNABojXilIDh15azsVM3sMZA5xPldCUHayXoUBmwXZdQ5HKS3CZrIqGnpmVnYuXgEhUXFJGVAsJM-ZAc0MCkun0fAmOkQykgGBM0DMsGgE0sxzkTmgJFwuAAFBxoHxIMhQFiAJSgAC8AD5QABvXCEZBaZQDeHkzjCDj+UAKKSwZA2LEAckCwVCIDWsAUHFQJEglgAVhxgJLUALiSRbNMsViAVopKTKTS6dsGTAMPISA5HFiGDqgSCDNhGBNYHx1FIMApYYDtHoDMSANxGwiqWCIeBWW3aOXQLAkwOEAC+xKNJpj5st1uZ5mwDAmWdgAdwSbjQA

You can also apply the rule https://app.requestly.io/rules/#sharedList/1670926548365-Regression-Repo and visit https://catfact.ninja/fact directly.

**Comment by lazyvab** [comment]
@mdpratt The example on https://catfact.ninja/fact didn't work for you because the static response in the shared rule was invalid. 
Instead of 
```js
{
    fact: "Mocked fact",
}
```
it should be:
```js
{
    "fact": "Mocked fact"
}
```
(JSON keys should be enclosed in double quotes and no trailing comma)

For `ArrayBuffer` case, it should not have altered the response type. It was found to be a bug on our side. We will roll out the fix soon.

However, Requestly currently supports modifying only JSON and text responses. It does not allow modifying `ArrayBuffer` responses yet.

**Comment by lazyvab** [comment]
Logged https://github.com/requestly/requestly/issues/253 to support modifying ArrayBuffer responses received in XHR.

**Comment by mdpratt** [comment]
Thanks for the fast turnaround. In theory, if the fix is made to not mutate the response type, then it should support JSON responses again again - even if it's not officially supported.

**Comment by lazyvab** [comment]
@mdpratt The issue has been fixed and rolled out in version `22.12.27`.
Please update your extension by either restarting the Chrome browser or clicking `Update` button on chrome://extensions/ page.

Requestly is not going to alter the `responseType` and will ignore the custom response in static mode if `responseType` is other than `text` or `json`. However, you may still be able to use Programmatic mode to override response and return a buffer as illustrated in the above comment.

Can you please check and confirm on your end, so we can close the ticket?
The other enhancement is being tracked under https://github.com/requestly/requestly/issues/253.

**Comment by mdpratt** [comment]
Hi @lazyvab, the fix does appear to be working for me, in that I can modify the response. However, if I change the response status code now (e.g to `500`), then that is now not being honoured.

**Comment by lazyvab** [comment]
Thanks @mdpratt for the confirmation. We are looking into the status code issue.

**Comment by mdpratt** [comment]
Hi @lazyvab, we're still not able to use `Modify API Response Rule` to set the status code. Do you want me to resolve this issue, and create a new one?

**Comment by lazyvab** [comment]
Hi @mdpratt. Sorry, we could not prioritize this last week. But we are urgently looking into this now and will resolve as earliest as possible. 
Let's keep this ticket open and I will let you know on this thread itself. Thanks!

**Comment by mdpratt** [comment]
I am able to change the response, however when setting the response status code to another value (ex. `500` or `404`) it does not have any effect; error handling logic is still seeing it as `200`.

![image](https://user-images.githubusercontent.com/982135/213580926-952984b7-9f8b-4402-8f6c-b0cc85dbeafa.png)

Exported rule (with IDs trimmed):

```json
[
  {
    "name": "404 SV1 - 23.1.14",
    "objectType": "rule",
    "pairs": [
      {
        "response": {
          "statusCode": "404",
          "statusText": "Not Found",
          "type": "code",
          "value": "function modifyResponse(args) {\n  return args.response\n}\n"
        },
        "source": {
          "filters": [],
          "key": "Url",
          "operator": "Equals",
          "value": "https://braket.us-west-2.amazonaws.com/device/arn%3Aaws%3Abraket%3A%3A%3Adevice%2Fquantum-simulator%2Famazon%2Fsv1"
        }
      }
    ],
    "ruleType": "Response",
    "status": "Active"
  }
]
```

**Comment by lazyvab** [comment]
@mdpratt Thanks for sharing the details. It was really helpful!

So, I tried on an AWS website and here is my observation:
An XMLHttpRequest (XHR) client goes through different states when a request is made. ([What does it mean?](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState))

Requestly currently overrides response and status code when the request is complete, i.e., XHR's ready state is 4 (DONE).
Whereas, the AWS code seems to read the status code when the request headers are received i.e. XHR's ready state is 2.

If you run below code in browser console:
```js
xhr = new XMLHttpRequest();
xhr.onreadystatechange = () => {
    if (xhr.readyState === XMLHttpRequest.DONE) {
        console.log("Status code = " + xhr.status);
    }
};
xhr.open("GET", "https://braket.us-west-2.amazonaws.com/device/arn%3Aaws%3Abraket%3A%3A%3Adevice%2Fquantum-simulator%2Famazon%2Fsv1");
xhr.send();
```
you would notice that it prints: `Status code = 404`.

I would suggest logging a different issue for this as it deviates from the original issue.

**Comment by lazyvab** [comment]
@mdpratt I have logged a separate ticket: https://github.com/requestly/requestly/issues/279 to override the status code immediately after response headers are received. We will provide an update when it is released. Thanks!

**Comment by lazyvab** [comment]
@mdpratt Meanwhile we fix it in the extension, you may unblock yourself using our [Desktop app](https://requestly.io/desktop). Once you login in the desktop application, all your browser rules should be synchronized. You could launch the browser from `Connected Apps` for testing.

**Comment by mdpratt** [comment]
Thanks @lazyvab, this is greatly appreciated. I'll make use of the desktop app for more demanding requests. I think this ticket can probably be closed now, as the original ask has been solved, and 2 new feature requests have been created.

**Comment by lazyvab** [comment]
Thanks @mdpratt. Closing this one now. We will actively track the other two feature requests.
